### DISCUSSION TITLE
Country-specific BBAN checks

### DISCUSSION THREAD

**Discussion by mhemeryck** [mdomke/schwifty#49] [Q&A]
Hi all, I have question related to country-specific BBAN checks.

It seems this library does not implement those checks?

Given that I like the developer API and the overall setup of this project, I would be interested to know if and how we could implement such a feature.

In broad terms, I am thinking of a custom exception (`invalidBBAN`?) when doing the validation of the IBAN.

**Comment by mdomke** [comment]
Hey @mhemeryck, sorry for the late reply. I think one would have to extend the `IBAN.validate()`-[method](https://github.com/mdomke/schwifty/blob/master/schwifty/iban.py#L193-L212), so that it calls a method like `IBAN._validate_bban()`. In that method one would select the appropriate checksum algorithm based on the country code. If this check fails I would raise an `InvalidBBAN`-exception. Maybe it is worth thinking about making this additional check optional e.g. by adding a parameter like `allow_invalid_bban=True` to the `IBAN`-constructor.

**Reply by mdomke** [reply]
The details of the BBAN validation might be a little bit more complicated depending on the country. I had a brief look at how this is done for German banks and there are over 100 checksum algorithms that can be selected depending on the domestic bank code.

**Reply by mhemeryck** [reply]
Hi @mdomke, thx for your reply!

The overall approach makes sense to me, but since each country has its own quirks wrt the BBAN validation, I would only implement it for our purposes (Belgian). This brings its own questions:

* first of all: would you be OK with you?
* how would the BBAN validation work for those countries where it's not available? pass silently (basically mimicking the current behavior) or some `NotImplementedError` (the first option would seem the most sane)
* where would you put country-specific checks?
* I suppose you also want to augment the IBAN generators such that these also match the new BBAN validation? (I did notice something specific for Italian BBANs somewhere in the code base).

**Reply by mdomke** [reply]
Let me make a proposal on how to do this in general by using German banks as an example and then we can discuss and see how additional country-specific implementations can be added to it. Sounds good?

**Reply by mhemeryck** [reply]
That would actually be really great! Thanks a lot already!

**Reply by mdomke** [reply]
I added support for country specific BBAN checksums in #50. This currently includes the account code checksum verification for German banks and the BBAN checksum verification for Italian banks. If you wanted to add algorithms for Belgian banks have a look at the abstract base class `Algorithm` in `schwifty.checksum`. It has two methods `compute` and `validate` which have to be implemented by algorithms that should be added to the "algorithms library". Each algorithm class has a `name` class-variable which is used for identification. The `schwifty.checksum.register`-function should be used to add an algorithm to the library by providing the country code (e.g. `"BE"`) as a prefix. In Germany there are over 150 different algorithms from which banks can chose. I adapted the german bank registry file so that each entry maps to the respective algorithm name. I don't know if you would need something similar in your case as well. Have a look at the `IBAN._validate_bban_checksum`-method to see how the actual algorithm selection is done. You can also have a look at `schwifty.checksum.germany` as inspiration on how to implement those algorithm classes. If you have any further questions don't hesitate to ask!

**Reply by mhemeryck** [reply]
Thx a lot for all the work!

I also really like the new API you added there for the BBAN checksum validation.

Fortunately, for the case of Belgium, it's a bit easier than for Germany (we only have 1 algorithm as I understand it).

See #51 for my implementation.

**Reply by mdomke** [reply]
@mhemeryck thank you very much for your contribution! I created a new release `2021.05.2` which includes your changes.

**Reply by mhemeryck** [reply]
Thanks a lot!

My contribution was fairly limited since it could readily build on the new API you did create :slightly_smiling_face:
